### PR TITLE
Few-spike neurons

### PR DESCRIPTION
**Pull request by neworderofjamie**
This is my initial attempt at integrating these neurons into mlGeNN. To do this I've refactored a few small bits, most significantly I've added 'converter' classes which encapsulate the mapping of TF activation types to mlGeNN neuron classes. Because normalisation is tied to a particular form of model conversion (e.g. with Few Spike, all you care about is maximum activations of each layer) this functionality is now incorporated into the converter classes. I didn't make any particular efforts to keep this backward compatible - it would be possible to fix up ``Model.convert_tf_model`` but, at the point in development, I think breaking things is fine.

Outstanding unpleasantness I'm not really sure what to do about (@jamesturner246 your thoughts would be much appreciated here):
- The automatic pipeline-depth calculation is a bit sketchy and definitely only works with sequential models
- The user needs to know to set ``time=K`` when they call ``Model.evaluate``. This _could_ be implemented via the converter class i.e. ``RateBased`` could take time as a constructor parameter and ``FewSpike`` could just return K **but**, if you defined a model in mlGeNN (rather than converting one from TF) this wouldn't work so maybe each neuron object should have a "required evaluation time" of something (in the style of the pipeline-depth stuff)

Less problematic things for another day:
- Currently only support ReLU neurons - could be trivially extended to generic form of FS
- When we split out the TF dependency, the converter classes will probably move to the same place.

I've updated the VGG16 and Simple CNN examples to take the ``--few-spike`` command line argument to trigger this mode of operation. In terms of throughput, the FS neuron-based conversion is between 60x and 95x faster (depending on batch size) than the rate-based:
VGG16 CIFAR10 - rate-based vs FS vs TF
SimpleCNN MNIST - rate-based vs FS vs TF
It's probably not worth getting very excited about mlGeNN beating TF in the simple CNN because I suspect the evaluate call which I'm timing also does some initialisation.

### CODE REVIEW

File: ml_genn/layers/fs_neurons.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    const int kInt = (int)$(K);
+
+    // Get timestep within presentation
+    const int pipeTimestep = (int)($(t) / DT);
+    const int phaseTimestep = pipeTimestep % kInt;
+
+    // Calculate magic constants. For RelU hT=h=T
+    // **NOTE** d uses last timestep as that was when spike was SENT
+    const scalar hT = $(scale) * (1 << (kInt - (1 + phaseTimestep)));
+    const scalar d = $(scale) * (1 << ((kInt - phaseTimestep) % kInt));
```
Comment (tnowotny) [inline ml_genn/layers/fs_neurons.py:21]: phaseTimestep < kInt, so the "% kInt" does nothing?

File: ml_genn/layers/fs_neurons.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
```
Comment (tnowotny) [inline ml_genn/layers/fs_neurons.py:58]: as above ... "% kInt" appears to be doing nothing

File: ml_genn/layers/fs_neurons.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    ''',
+    reset_code='''
+    $(Vmem) -= hT;
+    ''',
+    is_auto_refractory_required=False)
+    
+class FSReluNeurons(Neurons):
+    pipeline_stages = 0.5
+
+    def __init__(self, first_phase, K=10, alpha=25):
```
Comment (tnowotny) [inline ml_genn/layers/fs_neurons.py:83]: ... as above ... default k value a power of 2?

**Review by tnowotny — COMMENTED** [review]
I have a few comments/suggestions/questions ...

File: examples/simple_cnn.py [hunk 4]
```diff
@@ -43,20 +43,23 @@
     tf_model.evaluate(x_test, y_test)
 
     # Create, normalise and evaluate ML GeNN model
-    mlg_model = Model.convert_tf_model(tf_model, input_type=args.input_type, connectivity_type=args.connectivity_type)
+    converter = FewSpike() if args.few_spike else RateBased(args.input_type)
```
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply examples/simple_cnn.py:46]: Why do you think powers of 2 are more sensible? K=10 was the value they used in Figure 4 when they showed the ReLU function and it works for small models

[the remark this answers is not in the document]
Comment (tnowotny) [reply examples/simple_cnn.py:46]: seems more natural as it looks like a binary representation of a number where we are used to 8 bit chars, 16 bit shorts etc?

File: ml_genn/layers/fs_input_neurons.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    var_name_types=[('input', 'scalar', VarAccess_READ_ONLY_DUPLICATE), ('Vmem', 'scalar')],
+    sim_code='''
+    // Convert K to integer
+    const int kInt = (int)$(K);
+    
+    // Get timestep within presentation
+    const int pipeTimestep = (int)($(t) / DT);
+
+    // If this is the first timestep, apply input
+    if(pipeTimestep == 0) {
```
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply ml_genn/layers/fs_input_neurons.py:18]: 1. Indeed, the input model is implicitly in the first "phase" so input is set at t=0 (as is standard in mlGeNN) and it emits spikes during the first k timesteps
2. mlGeNN resets the GeNN time after every stimuli presentation which means you don't need that extra modulus

[the remark this answers is not in the document]
Comment (tnowotny) [reply ml_genn/layers/fs_input_neurons.py:18]: I see - that makes sense then.

File: ml_genn/layers/fs_neurons.py (hunk 1 above)
Comment (neworderofjamie) [reply ml_genn/layers/fs_neurons.py:21]: I had to think about this again when you asked this...They only differ in the first timestep of the phase i.e. ``phaseTimestep=0`` when you want to be  using the ``d`` value for the last timestep of the previous phase. So, if ``kInt=10``, ``1 << (10 - 0) = 1024`` which incorrectly applies a very large scale whereas ``1 << ((kInt - 0) % 10) = 1`` which is correct

File: examples/simple_cnn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply examples/simple_cnn.py:46]: But, unlike on boring modern computer architectures, there's no reason to use a power of two number of bits!

[the remark this answers is not in the document]
Comment (tnowotny) [reply examples/simple_cnn.py:46]: true, but at least we know approximately what kind of resolution we are getting ;-)
However, I do not feel strongly about this.

**Review by tnowotny — APPROVED** [review]
Yeah ... I think this is now very good indeed. But as discussed "offline", it will be key to reduce the number of spikes if it is to become competitive in speed with traditional ANN solutions (or Wolfgang's tensorflow fudge of this). But very good for now and so much better than rate based encodings.

**Review by jamesturner246 — APPROVED** [review]
It looks good and runs very fast compared to the rate based conversions. Definitely the best to have the conversion function call compile and post_compile_norm, rather than leaving it to the user. Only nitpick is the example instructions in the top comment of ml_genn/model.py are now presumably no longer working. Else everything seems fine.

**Pull request merged**